File: 0343-Integer_Break/Integer_Break.java
class Solution {
    HashMap<Integer,Integer> maxProduct = new HashMap<>();
    
    public int integerBreak(int n) {
        int max = 0;
        for(int i = 2; i <= n; i++){
            max = getLargestProduct(i);
        }
        return max;        
    }
    
    private int getLargestProduct(int n){
        if(n <= 2) return 1;
        if(maxProduct.containsKey(n)) return maxProduct.get(n);
        
        int currMax = 0;
        for(int i = 1; i <= n/2; i++){
            // 2 things to be checked - Stop breaking it further, breaking it further
            currMax = Math.max(currMax, i * getLargestProduct(n - i));
            currMax = Math.max(currMax, i * (n - i));
        }
        
        maxProduct.put(n, currMax);
        return currMax;
    }
}

File: 0343-IntegerBreak/IntegerBreak.go
package integerbreak

func integerBreak(n int) int {
    hm := make(map[int]int)
    
    max := 0
    for i := 2; i <= n; i++ {
        max = getLargestProduct(i, hm)
    }
    
    return max
}

func getLargestProduct(n int, hm map[int]int) int {
    if n <= 2 {
        return 1
    } 
    
    if val, exists := hm[n]; exists {
        return val;
    }
    
    currMax := 0
    for i := 1; i <= n/2; i++ {
        currMax = maxInt(currMax, i * getLargestProduct(n - i, hm))
        currMax = maxInt(currMax, i * (n - i))
    }
    
    hm[n] = currMax
    return currMax    
}


func maxInt(a, b int) int {
    if a > b {
        return a
    }
    return b
}